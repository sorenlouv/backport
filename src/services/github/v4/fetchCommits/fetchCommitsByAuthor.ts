import gql from 'graphql-tag';
import { isEmpty, uniqBy, orderBy } from 'lodash';
import { ValidConfigOptions } from '../../../../options/options';
import { filterNil } from '../../../../utils/filterEmpty';
import { filterUnmergedCommits } from '../../../../utils/filterUnmergedCommits';
import { HandledError } from '../../../HandledError';
import { swallowMissingConfigFileException } from '../../../remoteConfig';
import {
  Commit,
  SourceCommitWithTargetPullRequest,
  SourceCommitWithTargetPullRequestFragment,
  parseSourceCommit,
} from '../../../sourceCommit/parseSourceCommit';
import { apiRequestV4 } from '../apiRequestV4';
import { fetchAuthorId } from '../fetchAuthorId';

async function fetchByCommitPath({
  options,
  authorId,
  commitPath,
}: {
  options: {
    accessToken: string;
    githubApiBaseUrlV4?: string;
    maxNumber?: number;
    repoName: string;
    repoOwner: string;
    sourceBranch: string;
    dateSince: string | null;
    dateUntil: string | null;
  };
  authorId: string | null;
  commitPath: string | null;
}) {
  const {
    accessToken,
    githubApiBaseUrlV4 = 'https://api.github.com/graphql',
    maxNumber = 10,
    repoName,
    repoOwner,
    sourceBranch,
    dateSince,
    dateUntil,
  } = options;

  const query = gql`
    query CommitsByAuthor(
      $repoOwner: String!
      $repoName: String!
      $maxNumber: Int!
      $sourceBranch: String!
      $authorId: ID
      $commitPath: String
      $dateSince: GitTimestamp
      $dateUntil: GitTimestamp
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: $sourceBranch) {
          target {
            ... on Commit {
              history(
                first: $maxNumber
                author: { id: $authorId }
                path: $commitPath
                since: $dateSince
                until: $dateUntil
              ) {
                edges {
                  node {
                    ...SourceCommitWithTargetPullRequestFragment
                  }
                }
              }
            }
          }
        }
      }
    }

    ${SourceCommitWithTargetPullRequestFragment}
  `;

  const variables = {
    repoOwner,
    repoName,
    sourceBranch,
    maxNumber,
    authorId,
    commitPath,
    dateSince,
    dateUntil,
  };

  try {
    return await apiRequestV4<CommitByAuthorResponse>({
      githubApiBaseUrlV4,
      accessToken,
      query,
      variables,
    });
  } catch (e) {
    //@ts-expect-error
    return swallowMissingConfigFileException<CommitByAuthorResponse>(e);
  }
}

export async function fetchCommitsByAuthor(options: {
  accessToken: string;
  author: string | null;
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
  commitPaths?: string[];
  dateSince: string | null;
  dateUntil: string | null;
  githubApiBaseUrlV4?: string;
  maxNumber?: number;
  onlyMissing?: boolean;
  repoName: string;
  repoOwner: string;
  sourceBranch: string;
}): Promise<Commit[]> {
  const { sourceBranch, commitPaths = [] } = options;

  const authorId = await fetchAuthorId(options);
  const responses = await Promise.all(
    isEmpty(commitPaths)
      ? [fetchByCommitPath({ options, authorId, commitPath: null })]
      : commitPaths.map((commitPath) =>
          fetchByCommitPath({ options, authorId, commitPath })
        )
  );

  // we only need to check if the first item is `null` (if the first is `null` they all are)
  if (responses[0].repository.ref === null) {
    throw new HandledError(
      `The upstream branch "${sourceBranch}" does not exist. Try specifying a different branch with "--source-branch <your-branch>"`
    );
  }

  const commits = responses
    .flatMap((res) => {
      return res.repository.ref?.target.history.edges.map((edge) => {
        const sourceCommit = edge.node;
        return parseSourceCommit({ options, sourceCommit });
      });
    })
    .filter(filterNil);

  // terminate if not commits were found
  if (isEmpty(commits)) {
    const pathText =
      commitPaths.length > 0 ? ` touching files in path: "${commitPaths}"` : '';

    const errorText = options.author
      ? `There are no commits by "${options.author}" in this repository${pathText}. Try with \`--all\` for commits by all users or \`--author=<username>\` for commits from a specific user`
      : `There are no commits in this repository${pathText}`;

    throw new HandledError(errorText);
  }

  const commitsUnique = uniqBy(commits, (c) => c.sourceCommit.sha);
  const commitsSorted = orderBy(
    commitsUnique,
    (c) => c.sourceCommit.committedDate,
    'desc'
  );

  if (options.onlyMissing) {
    return commitsSorted.filter(filterUnmergedCommits);
  }

  return commitsSorted;
}

export interface CommitByAuthorResponse {
  repository: {
    ref: {
      target: {
        history: {
          edges: Array<{ node: SourceCommitWithTargetPullRequest }>;
        };
      };
    } | null;
  };
}
