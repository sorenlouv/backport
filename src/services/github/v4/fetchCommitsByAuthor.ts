import isEmpty from 'lodash.isempty';
import ora from 'ora';
import { BackportOptions } from '../../../options/options';
import { CommitChoice } from '../../../types/Commit';
import { filterEmpty } from '../../../utils/filterEmpty';
import { HandledError } from '../../HandledError';
import {
  getFirstCommitMessageLine,
  getFormattedCommitMessage,
} from '../commitFormatters';
import { apiRequestV4 } from './apiRequestV4';
import { fetchAuthorId } from './fetchAuthorId';
import { getTargetBranchesFromLabels } from './getTargetBranchesFromLabels';

export async function fetchCommitsByAuthor(
  options: BackportOptions
): Promise<CommitChoice[]> {
  const {
    accessToken,
    branchLabelMapping,
    githubApiBaseUrlV4,
    maxNumber,
    path,
    repoName,
    repoOwner,
    sourceBranch,
  } = options;

  const query = /* GraphQL */ `
    query CommitsByAuthor(
      $repoOwner: String!
      $repoName: String!
      $maxNumber: Int!
      $sourceBranch: String!
      $authorId: ID
      $historyPath: String
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: $sourceBranch) {
          target {
            ... on Commit {
              history(
                first: $maxNumber
                author: { id: $authorId }
                path: $historyPath
              ) {
                edges {
                  node {
                    oid
                    message
                    associatedPullRequests(first: 1) {
                      edges {
                        node {
                          repository {
                            owner {
                              login
                            }
                            name
                          }
                          number
                          mergeCommit {
                            oid
                          }
                          labels(first: 50) {
                            nodes {
                              name
                            }
                          }
                          timelineItems(
                            last: 20
                            itemTypes: CROSS_REFERENCED_EVENT
                          ) {
                            edges {
                              node {
                                ... on CrossReferencedEvent {
                                  source {
                                    __typename
                                    ... on PullRequest {
                                      title
                                      state
                                      baseRefName
                                      commits(first: 20) {
                                        edges {
                                          node {
                                            commit {
                                              message
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  `;

  const spinner = ora(
    `Loading commits from branch "${sourceBranch}"...`
  ).start();
  let res: DataResponse;
  try {
    const authorId = await fetchAuthorId(options);
    res = await apiRequestV4<DataResponse>({
      githubApiBaseUrlV4,
      accessToken,
      query,
      variables: {
        repoOwner,
        repoName,
        sourceBranch,
        maxNumber,
        authorId,
        historyPath: path || null,
      },
    });
    spinner.stop();
  } catch (e) {
    spinner.fail();
    throw e;
  }

  if (res.repository.ref === null) {
    throw new HandledError(
      `The upstream branch "${sourceBranch}" does not exist. Try specifying a different branch with "--source-branch <your-branch>"`
    );
  }

  const commits = res.repository.ref.target.history.edges.map((edge) => {
    // it is assumed that there can only be a single PR associated with a commit
    // that assumption might not hold true forever but for now it works out
    const pullRequestEdge = edge.node.associatedPullRequests.edges[0];
    const commitMessage = edge.node.message;
    const sha = edge.node.oid;

    // get the source pull request unless the commit was merged directly
    const sourcePullRequest = getSourcePullRequest({
      pullRequestEdge,
      options,
      sha,
    });

    // find any existing target pull requests
    const existingTargetPullRequests = getExistingTargetPullRequests(
      commitMessage,
      sourcePullRequest
    );

    const pullNumber =
      sourcePullRequest?.node.number || getPullNumberFromMessage(commitMessage);

    const formattedMessage = getFormattedCommitMessage({
      message: commitMessage,
      pullNumber,
      sha,
    });

    const labels = sourcePullRequest?.node.labels.nodes.map(
      (node) => node.name
    );
    const targetBranchesFromLabels = getTargetBranchesFromLabels({
      labels,
      branchLabelMapping,
    });

    return {
      sourceBranch,
      targetBranchesFromLabels,
      sha,
      formattedMessage,
      pullNumber,
      existingTargetPullRequests,
    };
  });

  // terminate if not commits were found
  if (isEmpty(commits)) {
    const pathText = options.path
      ? ` touching files in path: "${options.path}"`
      : '';

    const errorText = options.all
      ? `There are no commits in this repository${pathText}`
      : `There are no commits by "${options.author}" in this repository${pathText}. Try with \`--all\` for commits by all users or \`--author=<username>\` for commits from a specific user`;

    throw new HandledError(errorText);
  }

  return commits;
}

function getPullNumberFromMessage(firstMessageLine: string) {
  const matches = firstMessageLine.match(/\(#(\d+)\)/);
  if (matches) {
    return parseInt(matches[1], 10);
  }
}

function getSourcePullRequest({
  pullRequestEdge,
  options,
  sha,
}: {
  pullRequestEdge: PullRequestEdge | undefined;
  options: BackportOptions;
  sha: string;
}) {
  if (
    pullRequestEdge?.node.repository.name === options.repoName &&
    pullRequestEdge.node.repository.owner.login === options.repoOwner &&
    pullRequestEdge.node.mergeCommit.oid === sha
  ) {
    return pullRequestEdge;
  }
}

export function getExistingTargetPullRequests(
  commitMessage: string,
  sourcePullRequest: PullRequestEdge | undefined
) {
  if (!sourcePullRequest) {
    return [];
  }

  const firstMessageLine = getFirstCommitMessageLine(commitMessage);
  return sourcePullRequest.node.timelineItems.edges
    .filter(filterEmpty)
    .filter((item) => {
      const { source } = item.node;

      const isPullRequest = source.__typename === 'PullRequest';
      const isMergedOrOpen =
        source.state === 'MERGED' || source.state === 'OPEN';

      if (!isPullRequest || !isMergedOrOpen) {
        return false;
      }

      const commitMatch = source.commits.edges.some((commit) => {
        return (
          getFirstCommitMessageLine(commit.node.commit.message) ===
          firstMessageLine
        );
      });

      const prTitleMatch = source.title.includes(firstMessageLine);
      const prNumberMatch = source.title.includes(
        sourcePullRequest.node.number.toString()
      );

      return commitMatch || (prTitleMatch && prNumberMatch);
    })
    .map((item) => {
      const { source } = item.node;
      return {
        branch: source.baseRefName,
        state: source.state,
      };
    });
}

export interface DataResponse {
  repository: {
    ref: {
      target: {
        history: {
          edges: HistoryEdge[];
        };
      };
    } | null;
  };
}

interface HistoryEdge {
  node: {
    oid: string;
    message: string;
    associatedPullRequests: {
      edges: PullRequestEdge[];
    };
  };
}

export interface PullRequestEdge {
  node: {
    number: number;
    mergeCommit: {
      oid: string;
    };
    labels: {
      nodes: {
        name: string;
      }[];
    };
    repository: {
      owner: {
        login: string;
      };
      name: string;
    };
    timelineItems: {
      edges: (TimelineItemEdge | null)[];
    };
  };
}

interface TimelineItemEdge {
  node: {
    source: {
      __typename: string;
      title: string;
      state: 'OPEN' | 'CLOSED' | 'MERGED';
      baseRefName: string;
      commits: {
        edges: CommitEdge[];
      };
    };
  };
}

interface CommitEdge {
  node: {
    commit: {
      message: string;
    };
  };
}
