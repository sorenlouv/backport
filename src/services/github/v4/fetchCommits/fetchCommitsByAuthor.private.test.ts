import { getDevAccessToken } from '../../../../test/private/getDevAccessToken';
import { Commit } from '../../../sourceCommit/parseSourceCommit';
import { fetchCommitsByAuthor } from './fetchCommitsByAuthor';

describe('fetchCommitsByAuthor', () => {
  let devAccessToken: string;

  beforeAll(async () => {
    devAccessToken = await getDevAccessToken();
  });

  describe('commitPaths', () => {
    const getOptions = () => ({
      accessToken: devAccessToken,
      author: 'sqren',
      historicalBranchLabelMappings: [],
      maxNumber: 10,
      repoName: 'repo-with-different-commit-paths',
      repoOwner: 'backport-org',
      sourceBranch: 'main',
    });

    const getCommitMessages = (commits: Commit[]) => {
      return commits.map((c) =>
        c.originalMessage.replace(/(\r\n|\n|\r)/gm, '')
      );
    };

    it('returns all commits', async () => {
      const commits = await fetchCommitsByAuthor({
        ...getOptions(),
        commitPaths: [] as Array<string>,
      });

      const commitMessages = getCommitMessages(commits);
      expect(commitMessages).toEqual([
        'Edit all lyrics',
        'Update .backportrc.json',
        'Edit "Take on me"',
        'Add backportrc.json',
        'Add "Bohemian Rhapsody""',
        'Add "99 Luftballons"',
        'Add "Take on me"',
      ]);
    });

    it('only returns commits related to "99-luftballons.txt"', async () => {
      const commits = await fetchCommitsByAuthor({
        ...getOptions(),
        commitPaths: ['lyrics/99-luftballons.txt'],
      });

      const commitMessages = getCommitMessages(commits);
      expect(commitMessages).toEqual([
        'Edit all lyrics',
        'Add "99 Luftballons"',
      ]);
    });

    it('only returns commits related to "take-on-me.txt"', async () => {
      const commits = await fetchCommitsByAuthor({
        ...getOptions(),
        commitPaths: ['lyrics/take-on-me.txt'],
      });

      const commitMessages = getCommitMessages(commits);
      expect(commitMessages).toEqual([
        'Edit all lyrics',
        'Edit "Take on me"',
        'Add "Take on me"',
      ]);
    });

    it('removes duplicates and order by `committedDate`', async () => {
      const commits = await fetchCommitsByAuthor({
        ...getOptions(),
        commitPaths: [
          'lyrics/99-luftballons.txt',
          'lyrics/take-on-me.txt',
          'lyrics/bohemian-rhapsody.txt',
        ],
      });

      const commitMessages = getCommitMessages(commits);
      expect(commitMessages).toEqual([
        'Edit all lyrics',
        'Edit "Take on me"',
        'Add "Bohemian Rhapsody""',
        'Add "99 Luftballons"',
        'Add "Take on me"',
      ]);
    });
  });

  describe('expectedTargetPullRequests', () => {
    let res: Awaited<ReturnType<typeof fetchCommitsByAuthor>>;
    beforeEach(async () => {
      res = await fetchCommitsByAuthor({
        accessToken: devAccessToken,
        commitPaths: [],
        historicalBranchLabelMappings: [],
        maxNumber: 10,
        repoName: 'backport-e2e',
        repoOwner: 'backport-org',
        sourceBranch: 'master',
        author: null,
      });
    });

    it('returns related OPEN PRs', async () => {
      const commitWithOpenPR = res.find((commit) => commit.pullNumber === 9);
      expect(commitWithOpenPR?.expectedTargetPullRequests).toEqual([
        {
          branch: '7.8',
          state: 'OPEN',
          number: 10,
          url: 'https://github.com/backport-org/backport-e2e/pull/10',
        },
      ]);
    });

    it('returns related MERGED PRs', async () => {
      const commitWithMergedPRs = res.find((commit) => commit.pullNumber === 5);
      expect(commitWithMergedPRs?.expectedTargetPullRequests).toEqual([
        {
          branch: '7.x',
          state: 'MERGED',
          number: 6,
          url: 'https://github.com/backport-org/backport-e2e/pull/6',
        },
        {
          branch: '7.8',
          state: 'MERGED',
          number: 7,
          url: 'https://github.com/backport-org/backport-e2e/pull/7',
        },
      ]);
    });

    it('returns empty if there are no related PRs', async () => {
      const commitWithoutPRs = res.find((commit) => commit.pullNumber === 8);
      expect(commitWithoutPRs?.expectedTargetPullRequests).toEqual([]);
    });
  });
});
