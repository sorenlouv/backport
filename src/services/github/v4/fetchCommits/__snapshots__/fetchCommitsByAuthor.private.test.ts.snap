// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitsByAuthor snapshot request/response makes the right queries: Query: AuthorId 1`] = `
"query AuthorId($author: String!) {
  user(login: $author) {
    id
  }
}"
`;

exports[`fetchCommitsByAuthor snapshot request/response makes the right queries: Query: CommitsByAuthor 1`] = `
"query CommitsByAuthor($repoOwner: String!, $repoName: String!, $maxNumber: Int!, $sourceBranch: String!, $authorId: ID, $commitPath: String, $dateSince: GitTimestamp, $dateUntil: GitTimestamp) {
  repository(owner: $repoOwner, name: $repoName) {
    ref(qualifiedName: $sourceBranch) {
      target {
        ... on Commit {
          history(
            first: $maxNumber
            author: {id: $authorId}
            path: $commitPath
            since: $dateSince
            until: $dateUntil
          ) {
            edges {
              node {
                ...SourceCommitWithTargetPullRequestFragment
              }
            }
          }
        }
      }
    }
  }
}

fragment SourceCommitWithTargetPullRequestFragment on Commit {
  repository {
    name
    owner {
      login
    }
  }
  sha: oid
  message
  committedDate
  author {
    name
    email
  }
  associatedPullRequests(first: 1) {
    edges {
      node {
        url
        number
        labels(first: 50) {
          nodes {
            name
          }
        }
        baseRefName
        mergeCommit {
          ...RemoteConfigHistoryFragment
          sha: oid
          message
        }
        timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
          edges {
            node {
              ... on CrossReferencedEvent {
                targetPullRequest: source {
                  __typename
                  ... on PullRequest {
                    targetMergeCommit: mergeCommit {
                      sha: oid
                      message
                    }
                    repository {
                      name
                      owner {
                        login
                      }
                    }
                    url
                    title
                    state
                    baseRefName
                    number
                    commits(first: 20) {
                      edges {
                        node {
                          targetCommit: commit {
                            message
                            sha: oid
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}

fragment RemoteConfigHistoryFragment on Commit {
  remoteConfigHistory: history(first: 1, path: \\".backportrc.json\\") {
    edges {
      remoteConfig: node {
        committedDate
        file(path: \\".backportrc.json\\") {
          ... on TreeEntry {
            object {
              ... on Blob {
                text
              }
            }
          }
        }
      }
    }
  }
}"
`;

exports[`fetchCommitsByAuthor snapshot request/response returns the correct response 1`] = `
Array [
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [
      Object {
        "branch": "8.0",
        "mergeCommit": Object {
          "message": "[APM] Add note about synthtrace to APM docs (#121633) (#121643)

Co-authored-by: Søren Louv-Jansen <soren.louv@elastic.co>",
          "sha": "842adfdeb5541b059231857522f9009771a46107",
        },
        "number": 121643,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/121643",
      },
    ],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-12-20T14:20:16Z",
      "message": "[APM] Add note about synthtrace to APM docs (#121633)",
      "sha": "d421ddcf6157150596581c7885afa3690cec6339",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "[APM] Add note about synthtrace to APM docs (#121633)",
        "sha": "d421ddcf6157150596581c7885afa3690cec6339",
      },
      "number": 121633,
      "url": "https://github.com/elastic/kibana/pull/121633",
    },
  },
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-12-08T20:03:55Z",
      "message": "[APM] Prefer service.name for logs correlation (#120694)",
      "sha": "b1b491959dab47aeb83c88ee2accb2db46d23793",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "[APM] Prefer service.name for logs correlation (#120694)",
        "sha": "b1b491959dab47aeb83c88ee2accb2db46d23793",
      },
      "number": 120694,
      "url": "https://github.com/elastic/kibana/pull/120694",
    },
  },
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [
      Object {
        "branch": "8.0",
        "mergeCommit": Object {
          "message": "[APM] Disable telemetry in agent config endpoint (#120106) (#120186)

Co-authored-by: Søren Louv-Jansen <soren.louv@elastic.co>",
          "sha": "33c132edf127aafe0886921e7451cbcdb3e68acd",
        },
        "number": 120186,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/120186",
      },
    ],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-12-01T23:26:48Z",
      "message": "[APM] Disable telemetry in agent config endpoint (#120106)",
      "sha": "b1bb4a93959f19a653b9cfb207a5c6acb6559482",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "[APM] Disable telemetry in agent config endpoint (#120106)",
        "sha": "b1bb4a93959f19a653b9cfb207a5c6acb6559482",
      },
      "number": 120106,
      "url": "https://github.com/elastic/kibana/pull/120106",
    },
  },
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [
      Object {
        "branch": "8.0",
        "mergeCommit": Object {
          "message": "[APM] Remove index_pattern.json and add custom field formatters (#119915) (#120012)

* [APM] Remove index_pattern.json and add custom field formatters

* Fix tests

* Fix tests

* Fix tutorial

Co-authored-by: Søren Louv-Jansen <soren.louv@elastic.co>",
          "sha": "4d5992dbcbfc097f94562894882ad80c2791aad0",
        },
        "number": 120012,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/120012",
      },
    ],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-11-30T18:47:26Z",
      "message": "[APM] Remove index_pattern.json and add custom field formatters (#119915)

* [APM] Remove index_pattern.json and add custom field formatters

* Fix tests

* Fix tests

* Fix tutorial",
      "sha": "434f6e6a88faf24dc1ea41f9f726db78e46355a7",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "[APM] Remove index_pattern.json and add custom field formatters (#119915)

* [APM] Remove index_pattern.json and add custom field formatters

* Fix tests

* Fix tests

* Fix tutorial",
        "sha": "434f6e6a88faf24dc1ea41f9f726db78e46355a7",
      },
      "number": 119915,
      "url": "https://github.com/elastic/kibana/pull/119915",
    },
  },
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [
      Object {
        "branch": "8.0",
        "mergeCommit": Object {
          "message": "[APM] Remove log-log descriptions from correlation charts (#119700) (#120178)",
          "sha": "0bd9d7f9fe713ad78330fd1aa6c713c9c5c2c8d6",
        },
        "number": 120178,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/120178",
      },
      Object {
        "branch": "7.16",
        "mergeCommit": Object {
          "message": "[APM] Remove log-log descriptions from correlation charts (#119700) (#120179)

# Conflicts:
#	x-pack/plugins/apm/public/components/app/correlations/failed_transactions_correlations.tsx
#	x-pack/plugins/apm/public/components/app/correlations/latency_correlations.tsx
#	x-pack/plugins/apm/public/components/app/transaction_details/distribution/index.tsx",
          "sha": "749d1cfc1c7c33beec8fee58046e3d70dbd43e68",
        },
        "number": 120179,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/120179",
      },
    ],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-11-29T15:12:27Z",
      "message": "[APM] Remove log-log descriptions from correlation charts (#119700)",
      "sha": "ae9d51b7fe3ee6f30d0d196c782e0dcabb7ac5ff",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "[APM] Remove log-log descriptions from correlation charts (#119700)",
        "sha": "ae9d51b7fe3ee6f30d0d196c782e0dcabb7ac5ff",
      },
      "number": 119700,
      "url": "https://github.com/elastic/kibana/pull/119700",
    },
  },
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [
      Object {
        "branch": "8.0",
        "mergeCommit": Object {
          "message": "[APM] Fix bug in documentation on \`span.destination\` metrics (#119789) (#119840)

Co-authored-by: Søren Louv-Jansen <soren.louv@elastic.co>",
          "sha": "72a88ee6ae55dfacec76cccf0ece7ea6f4bfcc42",
        },
        "number": 119840,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/119840",
      },
    ],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-11-29T13:40:01Z",
      "message": "[APM] Fix bug in documentation on \`span.destination\` metrics (#119789)",
      "sha": "43253ecafa1365eae45d31dd284d0a0bf273bc4d",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "[APM] Fix bug in documentation on \`span.destination\` metrics (#119789)",
        "sha": "43253ecafa1365eae45d31dd284d0a0bf273bc4d",
      },
      "number": 119789,
      "url": "https://github.com/elastic/kibana/pull/119789",
    },
  },
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [
      Object {
        "branch": "8.0",
        "mergeCommit": Object {
          "message": "[APM] Remove projections (#118327) (#118675)

* Delete errors projection

* Remove \`getMetricsProjection\`

* Remove \`getServiceNodesProjection\`

* Fix tests

Co-authored-by: Kibana Machine <42973632+kibanamachine@users.noreply.github.com>

Co-authored-by: Søren Louv-Jansen <soren.louv@elastic.co>",
          "sha": "03657221841ae4c13cc490963e3c2fbcbde0a78e",
        },
        "number": 118675,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/118675",
      },
    ],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-11-16T11:44:55Z",
      "message": "[APM] Remove projections (#118327)

* Delete errors projection

* Remove \`getMetricsProjection\`

* Remove \`getServiceNodesProjection\`

* Fix tests

Co-authored-by: Kibana Machine <42973632+kibanamachine@users.noreply.github.com>",
      "sha": "b48268dd0c42c9dc1fe310f7c0b3327117d066a5",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "[APM] Remove projections (#118327)

* Delete errors projection

* Remove \`getMetricsProjection\`

* Remove \`getServiceNodesProjection\`

* Fix tests

Co-authored-by: Kibana Machine <42973632+kibanamachine@users.noreply.github.com>",
        "sha": "b48268dd0c42c9dc1fe310f7c0b3327117d066a5",
      },
      "number": 118327,
      "url": "https://github.com/elastic/kibana/pull/118327",
    },
  },
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [
      Object {
        "branch": "8.0",
        "mergeCommit": Object {
          "message": "Update query_debugging_in_development_and_production.md (#118491) (#118631)

Co-authored-by: Søren Louv-Jansen <soren.louv@elastic.co>",
          "sha": "81f69039af290230b7b9b0a35924622a031ac572",
        },
        "number": 118631,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/118631",
      },
    ],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-11-15T22:54:40Z",
      "message": "Update query_debugging_in_development_and_production.md (#118491)",
      "sha": "e66377790605274d2610f45eed35c9d8febdb7b9",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "Update query_debugging_in_development_and_production.md (#118491)",
        "sha": "e66377790605274d2610f45eed35c9d8febdb7b9",
      },
      "number": 118491,
      "url": "https://github.com/elastic/kibana/pull/118491",
    },
  },
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [
      Object {
        "branch": "8.0",
        "mergeCommit": Object {
          "message": "[APM] Ignore unavailable indices for ml jobs (#117632) (#117872)

Co-authored-by: Søren Louv-Jansen <soren.louv@elastic.co>",
          "sha": "748db6969738ac626be83c80b0d1bed2abd50b70",
        },
        "number": 117872,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/117872",
      },
    ],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-11-08T15:12:52Z",
      "message": "[APM] Ignore unavailable indices for ml jobs (#117632)",
      "sha": "acc7bd9d97a2147f752942dbcf01d7fb2478234b",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "[APM] Ignore unavailable indices for ml jobs (#117632)",
        "sha": "acc7bd9d97a2147f752942dbcf01d7fb2478234b",
      },
      "number": 117632,
      "url": "https://github.com/elastic/kibana/pull/117632",
    },
  },
  Object {
    "author": Object {
      "email": "soren.louv@elastic.co",
      "name": "Søren Louv-Jansen",
    },
    "expectedTargetPullRequests": Array [
      Object {
        "branch": "8.0",
        "mergeCommit": Object {
          "message": "[APM] Document \`_inspect\` flag (#117635) (#117831)

Co-authored-by: Søren Louv-Jansen <soren.louv@elastic.co>",
          "sha": "276fd505bce0a2ed2163d32c1de70ecf2414ab30",
        },
        "number": 117831,
        "state": "MERGED",
        "url": "https://github.com/elastic/kibana/pull/117831",
      },
    ],
    "sourceBranch": "main",
    "sourceCommit": Object {
      "committedDate": "2021-11-08T11:52:20Z",
      "message": "[APM] Document \`_inspect\` flag (#117635)",
      "sha": "d13ee69bdac1deebfea9f02a29cab0f9a87f2f8e",
    },
    "sourcePullRequest": Object {
      "mergeCommit": Object {
        "message": "[APM] Document \`_inspect\` flag (#117635)",
        "sha": "d13ee69bdac1deebfea9f02a29cab0f9a87f2f8e",
      },
      "number": 117635,
      "url": "https://github.com/elastic/kibana/pull/117635",
    },
  },
]
`;
